bazel/version/rules.bzl

```diff
@@ -0,0 +1,144 @@
+
+The substituted keys, and any guidance on values:
+- `VERSION` (the version string for Carbon)
+- `BUILD_EMBED_LABEL` (value of --embed_label)
+- `BUILD_HOST` (the name of the host machine running the build)
+- `BUILD_USER` (the name of the user running the build)
+- `GIT_COMMIT_SHA` (output of `git parse-rev --short HEAD` or `unknown`)
+- `GIT_DIRTY_SUFFIX` (`.dirty` if dirty client state or `` if unknown)
+- `BUILD_TIMESTAMP` (the time of the build in seconds since the Unix Epoch)
+- `ATTRIBUTE` (an optional attribute to apply to definitions, defaults to empty)
+"""
+
+load(":compute_version.bzl", "compute_version")
+
+_STAMP_DOC = """
+Follows behavior of the common 'stamp' attributes on rules. Set to 1 or 0 to
+force stamping with actual build info on or off respectively, and to -1 to
+follow the value of the command line flag `--stamp`.
+"""
+
+def _is_exec_config(ctx):
+    """Detect if this is the exec configuration, previously known as "host".
+
+    Sadly, there is not yet a supported way to detect this so replicate the
+    hacks others currently use. Bazel issue:
+    https://github.com/bazelbuild/bazel/issues/14444
+    """
+    return "-exec" in ctx.bin_dir.path or "/host/" in ctx.bin_dir.path
+
+def _expand_version_build_info_impl(ctx):
+    """Generates a file from a template, substituting version and build info."""
+    inputs = [ctx.file.template]
+
+    # The substitutions provided and their default values.
+    substitutions = {
+        "BUILD_EMBED_LABEL": "unknown",
+        "BUILD_HOST": "unknown",
+        "BUILD_TIMESTAMP": "unknown",
+        "BUILD_USER": "unknown",
+        "GIT_COMMIT_SHA": "unknown",
+        "GIT_DIRTY_SUFFIX": "",
+        "MAKE_WEAK": "0",
+        "VERSION": compute_version(ctx),
+    }
+    substitutions.update(ctx.attr.substitutions)
+
+    arguments = [
+        "--template=" + ctx.file.template.path,
+        "--output=" + ctx.outputs.out.path,
+    ] + [
+        "--substitution=" + key + "=" + value
+        for key, value in substitutions.items()
+    ]
+
+    # We only want to allow stamping outside of the exec configuration.
+    if not _is_exec_config(ctx):
+        # Look at the attribute.
+        stamp = ctx.attr.stamp
+
+        # If requested, use the command line flag to select.
+        if stamp == -1:
+            # Set the default from `--stamp` / `--nostamp` command line flag,
+            # which we detect through a macro and `config_setting`, and pipe
+            # through an attribute.
+            stamp = 1 if ctx.attr.internal_stamp_flag_detect else 0
+
+        # Add the status files if stamping.
+        if stamp == 1:
+            inputs += [
+                ctx.info_file,
+                ctx.version_file,
+            ]
+            arguments += [
+                "--status-file=" + ctx.info_file.path,
+                "--status-file=" + ctx.version_file.path,
+            ]
+
+    ctx.actions.run(
```

Comment:
Yeah, I had tried and it didn't work well.

It expanded keys in more places than expected for me -- so for example a `VERSION` in a comment, etc.

But more importantly, it can't read the status files (and we can't read them here), and so for the status files will require having *some* python template expander already, and it seemed cleaner to just use it everywhere here to avoid two subtly different template syntaxes.

Reply:
I'm confused this, I don't see VERSION in a comment, maybe I'll poke at it separately... For now this seems fine though.

Reply:
It isn't in the current comment -- this was in a previous attempt that had a different structure.

The key thing for me is that the keys in `expand_template` from bazel are expanded undilemited. I wanted *something* to set them apart from just the string content.

That, combined with the need to read from a status file anyways, made it seem like I sohuld just use one templating system with python.

Reply:
FWIW, I think the intent with `expand_template` is you delimit it more yourself (e.g., `VERSION` -> `{VERSION}`). Sort of meh, but not sure there's a good solution for the status file (which also seems a little disappointing, TBH -- but maybe there's something generic that can be done)

Reply:
All the examples I found with the status file roll their own source generation step. =[

What I'd really like is maybe `#embed` and having the starlark generate some content that we then copy into the status file and read. But anyways, merging for now and can revisit if there are more clean ways in the future.

Reply:
Mostly I'd been hoping to find some bazel module that implemented the read. ;)